--- a/mount_root.c
+++ b/mount_root.c
@@ -116,6 +116,12 @@ static int
 done(int argc, char *argv[1])
 {
-	struct volume *v = volume_find("rootfs_data");
+	char dataparam[BUFLEN];
+	char *dataname = "rootfs_data";
+	struct volume *v;
+
+	if (get_var_from_file("/proc/cmdline", "fstools_overlay_name", dataparam, sizeof(dataparam)))
+		dataname = dataparam;
+	v = volume_find(dataname);
 
 	if (!v)
 		return -1;
--- a/factoryreset.c
+++ b/factoryreset.c
@@ -31,6 +31,7 @@
 
 #include "libfstools/libfstools.h"
 #include "libfstools/volume.h"
+#include "libfstools/common.h"
 
 static int jffs2_mark(struct volume *v);
 
@@ -102,6 +103,8 @@
 
 int main(int argc, char **argv)
 {
+	char dataparam[64];
+	char *dataname = "rootfs_data";
 	struct volume *v;
 	int ch, yes = 0, reset = 0, keep = 0;
 
@@ -136,9 +139,11 @@
 		return -1;
 	}
 
-	v = volume_find("rootfs_data");
+	if (get_var_from_file("/proc/cmdline", "fstools_overlay_name", dataparam, sizeof(dataparam)))
+		dataname = dataparam;
+	v = volume_find(dataname);
 	if (!v) {
-		ULOG_ERR("MTD partition 'rootfs_data' not found\n");
+		ULOG_ERR("MTD partition '%s' not found\n", dataname);
 		return -1;
 	}
 
